parsl/monitoring/db_manager.py

```diff
@@ -251,40 +253,48 @@ def start(self, priority_queue, node_queue, resource_queue):
         self._resource_queue_pull_thread.start()
 
         """
-        maintain a set to track the tasks that are already INSERTED into database
+        maintain a set to track the tasks that are already INSERTed into database
         to prevent race condition that the first resource message (indicate 'running' state)
-        arrives before the first task message.
-        If race condition happens, add to left_messages and operate them later
+        arrives before the first task message. In such a case, the resource table
+        primary key would be violated.
+        If that happens, the message will be added to deferred_resource_messages and processed later.
 
         """
-        inserted_tasks = set()
-        left_messages = {}
+        inserted_tasks = set()  # type: Set[object]
+
+        # for any task ID, we can defer exactly one message, which is the
+        # assumed-to-be-unique first message (with first message flag set).
+        # The code prior to this patch will discard previous message in
+        # the case of multiple messages to defer.
+        deferred_resource_messages = {}  # type: Dict[str, Any]
 
         while (not self._kill_event.is_set() or
                self.pending_priority_queue.qsize() != 0 or self.pending_resource_queue.qsize() != 0 or
                priority_queue.qsize() != 0 or resource_queue.qsize() != 0):
 
             """
-            WORKFLOW_INFO and TASK_INFO messages
+            WORKFLOW_INFO and TASK_INFO messages (i.e. priority messages)
 
             """
             logger.debug("""Checking STOP conditions: {}, {}, {}, {}, {}""".format(
                               self._kill_event.is_set(),
                               self.pending_priority_queue.qsize() != 0, self.pending_resource_queue.qsize() != 0,
                               priority_queue.qsize() != 0, resource_queue.qsize() != 0))
 
-            # This is the list of first resource messages indicating that task starts running
-            first_messages = []
+            # This is the list of resource messages which can be reprocessed as if they
+            # had just arrived because the corresponding first task message has been
```

Comment:
no I mean *because*.

The resource messages did arrive *before* the corresponding first task messages, and so were placed into the `deferred_resource_messages` queue.

At this point in the code, we have processed the corresponding first task message, and so *because* of that, the code can process the messages as if they had just arrived.